### ISSUE TITLE
Multiple statements + Streaming may need a mysql2 API adjustment

### ISSUE THREAD

**Issue by sodabrew** [brianmario/mysql2#600]
The `:streaming` flag controls whether the `query` method should call the mysqlclient C API's `use_result` or `store_result` for the first result.

When multiple results are enabled, this still happens for the first result, but there's no equivalent decision for subsequent results. The multiple statements API includes `next_result`, `store_result`, and `abandon_result!`.

Idea #1: Add `client.use_result` and expose the difference in behavior to the user.

Idea #2: Secretly call `mysql_use_result` from `client.store_result` when streaming was enabled on the original query.

Idea #3: Remove `Mysql2::Client#store_result` from the mysql2 gem API, hide the difference inside the client state, and have `Mysql2::Client#next_result` return a `Mysql2::Result` directly. I'm not sure where `abandon_result!` would need to be moved, however. Maybe `Mysql2::Result#abandon!` ?

**Comment by brianmario** [comment]
I kinda like #3 the best, even though it's the most complex (I think).

This next topic is probably best discussed in another issue but:

-------------------------------

Having spent a lot more time dealing with the wire protocol lately, I'm starting to lean toward just _always_ using `mysql_use_result` and leaving the whole "buffer everything locally" or "stream" decision up to the caller. If they want it all buffered, they can just iterate over all the results as fasts as you can. Then deal with processing. Otherwise, just iterate over each result and process them as you get them. This would mean the socket buffer would fill and stack up all the way up the network chain until the server's call to `write(2)` would eventually block. And I think that's all ok?

For what it's worth, `ActiveRecord` reads all the results into memory to build `ActiveRecord::Result` objects so there's no difference between us using `use_result` or `store_result`. But even still, for people using mysql2 directly, they can do what I suggested above if they want to emulate `store_result`.

-------------------------------

Would that change how we think about this API change? (which I think should happen anyway)

**Comment by sodabrew** [comment]
I'm imagining a benchmark to see the cost of store_result vs. use_result, where the default behavior on the mysql2 side would be pulling the entire result set into Ruby memory space.

Today's defaults:

* mysql2 calls mysql_query and then mysql_store_result
* libmysqlclient reads the result into its data structures
* mysql2 calls mysql_fetch_row to retrieve each row into the data structure in the Mysql2::Result

Proposed new defaults:

* mysql2 calls mysql_query and then mysql_use_result
* mysql2 calls mysql_fetch_row to retrieve each row into the Mysql2::Result

On paper, it looks like we're removing a whole pile of memory use, but actual performance could go any way given the network read loop, allocation pattern, etc etc.

**Comment by brianmario** [comment]
> I'm imagining a benchmark to see the cost of store_result vs. use_result, where the default behavior on the mysql2 side would be pulling the entire result set into Ruby memory space.

Yeah so what if mysql2 itself always uses `mysql_use_result` but depending on if the `:streaming` option is set (in Ruby) we could either build an in-memory (Ruby) array of the results ahead of time - before returning a `Mysql2::Result` to the caller - or return a `Mysql2::Result` immediately and stream them off the wire as they iterate with `#each`?

That may be what you just suggested, and if so then consider us on the same page now ;)

**Comment by sodabrew** [comment]
Yep, on the same page!

**Comment by sodabrew** [comment]
I did some benchmarking this morning and saw a negligible performance difference (+/- 0.1% so within margin of error) between streaming and non-streaming when enumerating all rows. This was on localhost. I will try these again on a slower link, to see if `mysql_store_result` is using larger packet sizes or other tricks for performance benefits in a higher latency network.

**Comment by ioquatix** [comment]
libmariadb has a better interface for streaming results using event driven I/O. FYI.
